test/unittest/dds/subscriber/DataReaderTests.cpp

```diff
@@ -12,152 +12,787 @@
+            reset_lengths(data_values, infos);
+
+            EXPECT_EQ(code, data_reader->take_next_instance(data_values, infos));
+            if (ReturnCode_t::RETCODE_OK == code)
+            {
+                send_data(data_values, infos);
+                data_reader->wait_for_unread_message(time_to_wait);
+            }
+            check_return_loan(code, data_reader, data_values, infos);
+            reset_lengths(data_values, infos);
+
+            check_instance_methods(instance_ok_code, instance_bad_code, instance_ok_code,
+                    data_reader, data_values, infos);
+        }
+
+        // Check read/take and variants without loan
+        {
+            FooSeq data_values(1u);
+            SampleInfoSeq infos(1u);
+
+            ReturnCode_t expected_return_loan_ret = code;
+            if (ReturnCode_t::RETCODE_OK == code)
+            {
+                // Even when read returns data, no loan will be performed
+                expected_return_loan_ret = ReturnCode_t::RETCODE_PRECONDITION_NOT_MET;
+            }
+
+            EXPECT_EQ(code, data_reader->read(data_values, infos));
+            check_return_loan(expected_return_loan_ret, data_reader, data_values, infos);
+            reset_lengths(data_values, infos);
+            EXPECT_EQ(code, data_reader->read_next_instance(data_values, infos));
+            check_return_loan(expected_return_loan_ret, data_reader, data_values, infos);
+            reset_lengths(data_values, infos);
+
+            EXPECT_EQ(code, data_reader->take(data_values, infos));
+            if (ReturnCode_t::RETCODE_OK == code)
+            {
+                send_data(data_values, infos);
+                data_reader->wait_for_unread_message(time_to_wait);
+            }
+            check_return_loan(expected_return_loan_ret, data_reader, data_values, infos);
+            reset_lengths(data_values, infos);
+
+            EXPECT_EQ(code, data_reader->take_next_instance(data_values, infos));
+            if (ReturnCode_t::RETCODE_OK == code)
+            {
+                send_data(data_values, infos);
+                data_reader->wait_for_unread_message(time_to_wait);
+            }
+            check_return_loan(expected_return_loan_ret, data_reader, data_values, infos);
+            reset_lengths(data_values, infos);
+
+            check_instance_methods(instance_ok_code, instance_bad_code, expected_return_loan_ret,
+                    data_reader, data_values, infos);
+        }
     }
 
-    void deleteData(
-            void* /*data*/) override
+    DomainParticipant* participant_ = nullptr;
+    Subscriber* subscriber_ = nullptr;
+    Publisher* publisher_ = nullptr;
+    Topic* topic_ = nullptr;
+    DataReader* data_reader_ = nullptr;
+    DataWriter* data_writer_ = nullptr;
+    TypeSupport type_;
+
+    InstanceHandle_t handle_ok_ = HANDLE_NIL;
+    InstanceHandle_t handle_wrong_ = HANDLE_NIL;
+
+};
+
+TEST_F(DataReaderTests, read_take_apis)
```

Comment:
It would be good to have a commentary explaining each of the tests as it is done [here](https://github.com/eProsima/Fast-DDS/blob/master/test/unittest/logging/LogTests.cpp#L337-L418).

Reply:
Done in b34942a66b1d823a99d6853ae87ca080a2696f69